File: 프로그래머스/2/12978. 배달/배달.py
import heapq

def solution(N, road, K):
    answer = 0

    # k 시간 이하로 배달할 수 있는 곳의 갯수 return
    # 1번에서 시작 -> 최단 경로 문제
    # road 에서 [마을1, 마을2, 시간]
    # 가중치 있음 -> 다익스트라. 없으면 BFS
    
    INF = int(1e9)
    dist = [INF] * (N+1) # 최단거리 저장용
    dist[1] = 0
    pq = []
    heapq.heappush(pq, (1, 0)) # 현재 노드, 현재 시간
    graph = [[] for _ in range(N+1)]
    
    for node in road:
        a = node[0]
        b = node[1]
        cost = node[2]
        
        graph[a].append((b,cost))
        graph[b].append((a,cost))
        
    
    while pq:
        node, time = heapq.heappop(pq)
        
        for info in graph[node]:
            nextNode, cost = info
            
            if dist[nextNode] < time + cost:
                # 이미 작음
                continue
            
            dist[nextNode] = cost + time
            heapq.heappush(pq, (nextNode, time + cost))
            
    print(dist)
        
    for item in dist:
        if item <= K:
            answer += 1
    
    
    return answer

File: 프로그래머스/2/12978. 배달/배달.js
function solution(N, road, K) {
    var answer = 0;

    // 1번에서 k시간 이하로 배달 가능한 마을 -> 다익스트라
    
    // 그래프 초기화
    const graph = Array.from({length : N+1}, () => Array(N+1).fill(Infinity))
    for (const [x,y,time] of road) {
        graph[x][y] = Math.min(time,graph[x][y])
        graph[y][x] = Math.min(time,graph[y][x])

    }
    
    const dist = Array(N+1).fill(Infinity) // 최단거리 배열
    const visit = Array(N+1).fill(0) // 방문 배열
   
    dist[1] = 0
    
    // N개의 마을을 돌면서 확정
    for (let i=1; i<= N; i++) {
        let current = -1
        let minDist = Infinity
        
        // 현재 방문하지 않은 노드 중에서 최단거리가 가장 작은 노드 찾기
        for (let k=1; k<=N; k++) {
            if (visit[k] !== 1 && dist[k] < minDist) {
                current = k
                minDist = dist[k]
            }
        }
        
        // 다 방문함
        if (current === -1) {
            break
        }
        
        visit[current] = 1
        
        // 노드와 연결된 모드 노드 거리 갱신
        for (let next=1; next<=N; next++) {
            
            if (visit[next] === 1) continue
            if (graph[current][next] === Infinity) continue // 연결 X
            
            const nextDist = dist[current] + graph[current][next] 
            
            // 갱신
            if (nextDist < dist[next]) {
                dist[next] = nextDist
            }
     
        }
        
        
    }
    
    for (const item of dist) {
        if (item <= K) {
            answer += 1
        } 
    }
    

    return answer;
}